src/main/java/de/wirvsvirus/hack/service/IntervalIntersectionUtil.java

```diff
@@ -0,0 +1,115 @@
+      final List<Pair<Instant, Instant>> list,
+      final Instant truncateTimestamp) {
+
+    return intersectList(list, Collections.singletonList(
+        Pair.of(INSTANT_MIN, truncateTimestamp)));
+  }
+
+  public static List<Pair<Instant, Instant>> intersectList(
+      final List<Pair<Instant, Instant>> list1,
+      final List<Pair<Instant, Instant>> list2) {
+    final List<Pair<Long, Long>> epochStyle = IntervalIntersectionUtil.intersectListOfLongs(
+        list1.stream()
+            .map(p -> Pair.of(p.getLeft().toEpochMilli(), p.getRight().toEpochMilli()))
+            .collect(Collectors.toList()),
+        list2.stream()
+            .map(p -> Pair.of(p.getLeft().toEpochMilli(), p.getRight().toEpochMilli()))
+            .collect(Collectors.toList())
+    );
+
+    return
+        epochStyle.stream()
+            .map(
+                p -> Pair.of(Instant.ofEpochMilli(p.getLeft()), Instant.ofEpochMilli(p.getRight())))
+            .collect(Collectors.toList());
+  }
+
+  static List<Pair<Long, Long>> intersectListOfLongs(
+      final List<Pair<Long, Long>> list1,
+      final List<Pair<Long, Long>> list2) {
+    checkRangesListPreconditions(list1);
+    checkRangesListPreconditions(list2);
+
+    final List<Pair<Long, Long>> output = new ArrayList<>();
+
+    // i and j pointers for arr1 and
+    // arr2 respectively
+    int i = 0;
+    int j = 0;
+
+    final int n = list1.size();
+    final int m = list2.size();
+
+    // Loop through all intervals unless
+    // one of the interval gets exhausted
+    while (i < n && j < m) {
+
+      // Left bound for intersecting segment
+      final long l = Math.max(list1.get(i).getLeft(), list2.get(j).getLeft());
+
+      // Right bound for intersecting segment
+      final long r = Math.min(list1.get(i).getRight(), list2.get(j).getRight());
+
+      // If segment is valid print it
+      if (l < r) {
+        output.add(Pair.of(l, r));
+      }
+
+      // If i-th interval's right bound is
+      // smaller increment i else increment j
+      if (list1.get(i).getRight() < list2.get(j).getRight()) {
+        i++;
+      } else {
+        j++;
+      }
+    }
+    return output;
+  }
+
+  static void checkRangesListPreconditions(final List<Pair<Long, Long>> list) {
+    StreamEx.of(list)
+        .forEach(
+            p -> Preconditions.checkArgument(
+                p.getLeft() <= p.getRight(),
+                "Range %s malformed", p));
+    final boolean strictMonotonicAsc = !StreamEx.of(list)
+        .pairMap((l, r) -> {
+          Preconditions.checkArgument(l.getLeft() < r.getLeft(), "Ranges must be ordered");
+          return l.getRight() <= r.getLeft();
+        })
+        .has(false);
+    if (!strictMonotonicAsc) {
+      throw new IllegalArgumentException("Ranges must be ordered strict monotonic!");
+    }
+    Preconditions.checkArgument(strictMonotonicAsc, "Ranges must be ordered strict monotonic!");
+  }
+
+  private IntervalIntersectionUtil() {
+  }
```

Comment:
WTF?

Reply:
https://stackoverflow.com/questions/14398747/hide-utility-class-constructor-utility-classes-should-not-have-a-public-or-def